### PR TITLE
CSharpScript should not own method infos of the base class

### PR DESCRIPTION
**Pull request by paulloz**
- Supersedes #77273

This is simply a rebased and squashed version of the original PR. We took way too long to review this, and the original author has not been answering the last ping. But this is still a needed, and appreciated, fix.

### CODE REVIEW

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs [hunk 1]
```diff
@@ -804,11 +804,11 @@ static void AppendTypeName(StringBuilder sb, Type type)
                 using var methods = new Collections.Array();
 
                 Type? top = scriptType;
```
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs:806]: This is now unused.

```suggestion
```

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs [hunk 2]
```diff
@@ -804,11 +804,11 @@ static void AppendTypeName(StringBuilder sb, Type type)
...
-                Type native = GodotObject.InternalGetClassNativeBase(top);
+                Type native = GodotObject.InternalGetClassNativeBase(scriptType);
 
-                while (top != null && top != native)
+                if (scriptType != null && scriptType != native)
```
Comment (raulsntos) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Bridge/ScriptManagerBridge.cs:809]: `scriptType` should never be null, so I don't think we need to check.

```suggestion
                if (scriptType != native)
```

**Review by raulsntos — COMMENTED** [review]
Thanks. I think it makes sense for `CSharpScript` to only contain the members declared in the C# type that it represents, then the inherited members can be retrieved by iterating the hierarchy chain.

The same change should likely be made to the other members (properties and signals), and we must make sure the behavior matches GDScript (i.e. if `GDScript::get_method_list` iterates the hierarchy chain and returns inherited methods, then `CSharpScript::get_method_list` should return inherited methods too).

For example, we still have a loop in _RPC functions_ that retrieve the RPC attribute for methods from base types, and in _Event signals_ that retrieve the signals from base types.

The method generated method `GetGodotPropertyList` would similarly have to be updated to only consider the properties in the current type (and not the base types), this would also fix https://github.com/godotengine/godot/issues/75271.

I know this sounds like a lot, but I'd rather make all these changes in the same PR to avoid ending up in an partial change where some members are retrieved only for the current type, and others also retrieve from the base types.

**Comment by paulloz** [comment]
> I know this sounds like a lot, but I'd rather make all these changes in the same PR to avoid ending up in an partial change where some members are retrieved only for the current type, and others also retrieve from the base types.

While I share the sentiment in principle, this PR is someone else's work that I salvaged because we were saying it should and could still be merged in our internal discussions. I'm not sure that I'm comfortable changing the scope of the PR, and putting their name on code they did not initially intend.

**Comment by paulloz** [comment]
Currently, the way we connect signals relies on the fact scripts signal lists include all the hierarchy[^1]. I'm not entirely sure about RPC functions, because I'm not familiar with that part of the codebase at all.

[^1]: https://github.com/godotengine/godot/blob/c4279fe3e0b27d0f40857c00eece7324a967285f/modules/mono/csharp_script.cpp#L1774

**Comment by raulsntos** [comment]
> Currently, the way we connect signals relies on the fact scripts signal lists include all the hierarchy.

We should change that to iterate the hierarchy, I see that you've already done this in 1d7085571efbe7b47c496806d490dde8dd6ad430.

> I'm not entirely sure about RPC functions, because I'm not familiar with that part of the codebase at all.

It looks like the dictionary returned by `Script::get_rpc_config` must contain the RPC config inherited from base types, so I guess we can keep the current implementation.

For the reference, GDScript initializes this dictionary by duplicating the dictionary from the base script:

https://github.com/godotengine/godot/blob/1d7085571efbe7b47c496806d490dde8dd6ad430/modules/gdscript/gdscript_compiler.cpp#L2721-L2725

**Review by raulsntos — APPROVED** [review]
Looks good and works as expected, thanks!

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**